src/shims/unix/fd.rs

```diff
@@ -195,37 +206,122 @@ impl FileDescription for NullOutput {
+        Ref::map(self.0.file_description.borrow(), |fd| fd.as_ref())
     }
 
     pub fn borrow_mut(&self) -> RefMut<'_, dyn FileDescription> {
-        RefMut::map(self.0.borrow_mut(), |fd| fd.as_mut())
+        RefMut::map(self.0.file_description.borrow_mut(), |fd| fd.as_mut())
     }
 
-    pub fn close<'ctx>(self, communicate_allowed: bool) -> InterpResult<'ctx, io::Result<()>> {
+    pub fn close<'tcx>(
+        self,
+        communicate_allowed: bool,
+        ecx: &mut MiriInterpCx<'tcx>,
+    ) -> InterpResult<'tcx, io::Result<()>> {
         // Destroy this `Rc` using `into_inner` so we can call `close` instead of
         // implicitly running the destructor of the file description.
+        let id = self.get_id();
         match Rc::into_inner(self.0) {
-            Some(fd) => RefCell::into_inner(fd).close(communicate_allowed),
+            Some(fd) => {
+                // Remove entry from the global epoll_interest table.
+                ecx.machine.epoll_interests.remove(id);
+
+                RefCell::into_inner(fd.file_description).close(communicate_allowed, ecx)
+            }
             None => Ok(Ok(())),
         }
     }
+
+    pub fn downgrade(&self) -> WeakFileDescriptionRef {
+        WeakFileDescriptionRef { weak_ref: Rc::downgrade(&self.0), id: self.get_id() }
+    }
+
+    pub fn get_id(&self) -> FdID {
+        self.0.id
+    }
+
+    /// Function used to retrieve the readiness event status of a file description and insert
+    ///  an `EpollReturn` into the ready list if the file description is ready.
+    pub(crate) fn check_and_update_readiness<'tcx>(
+        &self,
+        ecx: &mut InterpCx<'tcx, MiriMachine<'tcx>>,
+    ) -> InterpResult<'tcx, ()> {
+        ecx.check_and_update_readiness(self.get_id(), |ecx| {
+            self.borrow_mut().get_epoll_ready_events(ecx)
+        })
+    }
+}
+
+/// WeakFileDescriptorRef holds a weak reference to the actual file description.
+#[derive(Clone, Debug, Default)]
+pub struct WeakFileDescriptionRef {
+    weak_ref: Weak<FileDescWithID<dyn FileDescription>>,
+    id: FdID,
+}
+
+impl WeakFileDescriptionRef {
+    pub fn upgrade(&self) -> Option<FileDescriptionRef> {
+        Some(FileDescriptionRef(self.weak_ref.upgrade()?))
+    }
+}
+
+// PartialEq, Eq, PartialOrd and Ord are implemented here because WeakFileDescriptionRef
+// is used as a key in the interest_list and ready_list in Epoll. It is ordered using
+// file description ID assigned.
+
+impl PartialEq for WeakFileDescriptionRef {
+    fn eq(&self, other: &Self) -> bool {
+        self.id.eq(&other.id)
+    }
+}
+
+impl Eq for WeakFileDescriptionRef {}
+
+impl PartialOrd for WeakFileDescriptionRef {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for WeakFileDescriptionRef {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.id.cmp(&other.id)
+    }
+}
+
+/// Wrapper struct for file description ID.
```

Comment:
The fact that it's a wrapper is not useful to users. This is the id for all intents and purposes.

```suggestion
/// A unique id for file descriptions. While we could use the address, considering that
/// is definitely unique, the address would expose interpreter internal state when used
/// for sorting things. So instead we generate a unique id per file description that stays
/// the same even if a file descriptor is duplicated and gets a new integer file descriptor.
```